### ISSUE TITLE
How to type composition of HOC?

### ISSUE THREAD

**Issue by michalkvasnicak** [facebook/flow#3422]
Is there a way to type `compose` function which can accept various number of HOCs?

### Example

```js
declare function injectDate<Props, C: React.Component<*, Props, *>>(
  Komponent: Class<C>
): Class<React.Component<void, $Diff<Props, {date: Date}>, void>>;

declare function injectTime<Props, C: React.Component<*, Props, *>>(
  Komponent: Class<C>
): Class<React.Component<void, $Diff<Props, {time: number}>, void>>;

compose(injectDate, injectTime): Class<React.Component<void, $Diff<Props, PropsFromInjectDate & PropsFromInjectTime>, void>>

const Component = compose(injectDate, injectTime)(Timeless);
```

**Comment by AriaMinaei** [comment]
This might help: https://github.com/nordsoftware/react-flow-types

**Comment by mwalkerwells** [comment]
There is definitely a way. Do you have a specific question I could try to answer?

**Comment by michalkvasnicak** [comment]
@AriaMinaei Thank you but this like did not help.

@mwalkerwells I need to type `compose` function which can accept arbitrary number of decorators as arguments, do array reduce internally and return only one decorator with intersection of prop types returned from decorators.

Now I have to do something like this in code, because I don't know how to type compose.

```js
export default withInjectDate(
  withInjectTime(
    withSomethingElse(MyComponent)
  )
);
```

I want something like this

```js
export default compose(withInjectDate, withInjectTime, withSomethingElse)(MyComponent);
```

This should accept `1..n` high order functions and return only one with intersection of props from all passed HOCs and pass through flow type check.

I searched for composition and I found that it is tricky to do but I didn't find any example for HOCs.
